### PR TITLE
feat(Resource Acces Management): Create resource access rule

### PR DESCRIPTION
**Pull request by dmyios**
## Description

Implementation of Create Ressource Access Rule functionnality (form and API call)

[screen-capture (3).webm](https://github.com/centreon/centreon/assets/108675430/f27520b3-078e-4a82-9517-5637aa79779a)

**Fixes** # MON-24793

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [x] 24.04.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by bdauria** [comment]
I think the trash buttons should be red (@Wedierre )

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/AddEditResourceAccessRuleModal.tsx [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import { Modal } from '@centreon/ui/components';
+
+import {
+  labelCreateResourceAccessRule,
+  labelEditResourceAccessRule
+} from '../translatedLabels';
+
+import useResourceAccessRuleModal from './useResourceAccessRuleModal';
+import useModalStyles from './Modal.styles';
+import Form from './Form/Form';
```
Comment (L0th0mat1c) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/AddEditResourceAccessRuleModal.tsx:16]: Maybe we can change the name of file from Form to index.tsx and we can import component like this: 
``import Form from './Form``. This gives more lisibility

**Comment by Wedierre** [comment]
> I think the trash buttons should be red (@Wedierre )

Yes we did a test with Dmytro in blue and after reflection, let’s go on the proto.

On the prototype we had the circle around the trash pictogram in grey and the overflight in red. Perhaps we could keep this functioning.
We would have a less visible and therefore less intrusive trash pictogram for the user who on the other hand at the hover before the alert click by switching to red.
Capture d’écran 2024-01-18 à 11 56 53

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/AddEditResourceAccessRuleModal.tsx (hunk 1 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/AddEditResourceAccessRuleModal.tsx]: Done

**Review by Wedierre — APPROVED** [review]

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/FormInputs.styles.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  resourceSelection: {
+    display: 'flex',
+    flexDirection: 'column',
+    justifyContent: 'flex-start',
+    width: '100%'
+  },
+  ruleProperties: {
+    backgroundColor: theme.palette.background.default,
+    borderRadius: '4px',
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/FormInputs.styles.ts:21]:
```suggestion
    borderRadius: theme.spacing(.5),
```

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        backgroundColor: theme.palette.success.main,
+        opacity: 1
+      },
+      color: theme.palette.common.white
+    }
+  },
+  '& .MuiSwitch-thumb': {
+    backgroundColor: theme.palette.common.white
+  }
+}));
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: move this to styles file

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilter.tsx [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                color: 'primary',
+                onDelete: (_, option): void =>
+                  deleteResourceItem({
+                    index: resourceIndex,
+                    option,
+                    resources: resource.resources
+                  })
+              }}
+              className={classes.resources}
+              dataTestId={`${labelSelectResource}`}
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilter.tsx:89]:
```suggestion
              dataTestId={labelSelectResource}
```

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    const lastDataset = last(lastDatasetFilter);
+    if (isNil(lastDataset)) {
+      return false;
+    }
+
+    return isEmpty(lastDataset.resourceType) || isEmpty(lastDataset.resources);
+  };
+
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:45]: this logic can be simplified. for example you can combine the three "if condition" into a single one
`    const lastDataset = last(lastDatasetFilter);

    if (isNil(datasetFiltersArray) || isNil(lastDataset) || isNil( last(lastDataset)) ) {
        return false;
    }`
You can further simplify the code based on the specific logic you intend to implement.

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  const { addDatasetFilter, datasetFilters, deleteDatasetFilter } =
+    useDatasetFilters();
+
+  const areResourcesFilled = (datasets: Array<Dataset>): boolean =>
+    datasets?.every(
+      ({ resourceType, resources }) =>
+        !isEmpty(resourceType) && !isEmpty(resources)
+    );
+
+  const areAddDatasetButtonDisabled = (
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:26]:
```suggestion
  const isAddDatasetButtonDisabled = (
```

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx (hunk 5 above)
Comment (Med1233) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:45]: is datasetFiltersArray can be null ?

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    if (isNil(lastDataset)) {
+      return false;
+    }
+
+    return isEmpty(lastDataset.resourceType) || isEmpty(lastDataset.resources);
+  };
+
+  return (
+    <div>
+      {datasetFilters.map((datasetFilter, index) => (
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:48]:
```suggestion
      {datasetFilters?.map((datasetFilter, index) => (
```

Comment (Med1233) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:48]: is datasetFiltersArray can be null ?

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/hooks/useDatasetFilter.ts [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+        { id: ResourceTypeEnum.ServiceCategory, name: labelServiceCategory }
+      ];
+    }
+
+    if (
+      equals(value[index - 1]?.resourceType, ResourceTypeEnum.MetaService) ||
+      equals(value[index - 1]?.resourceType, ResourceTypeEnum.Service)
+    ) {
+      return [];
+    }
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/hooks/useDatasetFilter.ts:179]: you can also simplify this. check what is done [here ](https://github.com/centreon/centreon/blob/develop/centreon/www/front_src/src/Dashboards/SingleInstancePage/Dashboard/AddEditWidget/WidgetProperties/Inputs/Resources/useResources.ts#L198)

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    color: theme.palette.common.white,
+    margin: '0px'
+  },
+  deleteIconChip: {
+    '&:hover': {
+      backgroundColor: theme.palette.chip.color.error
+    },
+    backgroundColor: theme.palette.divider,
+    position: 'relative',
+    right: theme.spacing(0.6)
```
Comment (Med1233) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts:26]:
```suggestion
    right: theme.spacing(0.5)
```

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx (hunk 3 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: Done

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx (hunk 6 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:26]: Done

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx (hunk 3 above)
Comment (Med1233) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: why the is not visible in the demo ? is not implemented yet ?

Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: It is visible, it's in Rule Properties section, under Description text area

Comment (Med1233) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: I meant in the Listing

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx (hunk 7 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/DatasetFilters.tsx:48]: No, it cannot

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx (hunk 3 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/components/ActivateSwitch.tsx:26]: It'll be implemented with PATCH endpoint

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts (hunk 9 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts:26]: image

It allows to align the chip with delete button on the brace

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/hooks/useDatasetFilter.ts (hunk 8 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/hooks/useDatasetFilter.ts]: done

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts (hunk 9 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/FormInputs/styles/DeleteDatasetButton.styles.ts]: Done

**Review by Med1233 — APPROVED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 10]
```diff
@@ -18593,3 +18593,36 @@ msgstr "Utilisateur supprimé"
...
+msgid "Edit resource access rule"
+msgstr "Modifier une règle d'accès aux ressources"
+
+msgid "Rule properties"
+msgstr "Propriétés d'une règle"
+
+msgid "Resource selection"
+msgstr "Sélection des ressources"
+
+msgid "Contacts / Contact groups"
```
Comment (Thebarda) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18609]:
```suggestion
msgid "Contacts/Contact groups"
```

**Review by Med1233 — DISMISSED** [review]

File: centreon/lang/pt_BR.UTF-8/LC_MESSAGES/messages.po [hunk 11]
```diff
@@ -17260,3 +17260,36 @@ msgstr ""
...
+# msgid "Edit resource access rule"
+# msgstr ""
+
+# msgid "Rule properties"
+# msgstr ""
+
+# msgid "Resource selection"
+# msgstr ""
+
+# msgid "Contacts / Contact groups"
```
Comment (Thebarda) [inline centreon/lang/pt_BR.UTF-8/LC_MESSAGES/messages.po:17276]:
```suggestion
# msgid "Contacts/Contact groups"
```

File: centreon/lang/pt_PT.UTF-8/LC_MESSAGES/messages.po [hunk 12]
```diff
@@ -17249,3 +17249,36 @@ msgstr "Você não salvou as alterações, deseja continuar?"
...
+# msgid "Edit resource access rule"
+# msgstr ""
+
+# msgid "Rule properties"
+# msgstr ""
+
+# msgid "Resource selection"
+# msgstr ""
+
+# msgid "Contacts / Contact groups"
```
Comment (Thebarda) [inline centreon/lang/pt_PT.UTF-8/LC_MESSAGES/messages.po:17265]:
```suggestion
# msgid "Contacts/Contact groups"
```

File: centreon/lang/de_DE.UTF-8/LC_MESSAGES/messages.po [hunk 13]
```diff
@@ -16466,3 +16466,36 @@ msgstr "Erweiterter Modus"
...
+# msgid "Edit resource access rule"
+# msgstr ""
+
+# msgid "Rule properties"
+# msgstr ""
+
+# msgid "Resource selection"
+# msgstr ""
+
+# msgid "Contacts / Contact groups"
```
Comment (Thebarda) [inline centreon/lang/de_DE.UTF-8/LC_MESSAGES/messages.po:16482]:
```suggestion
# msgid "Contacts/Contact groups"
```

File: centreon/lang/es_ES.UTF-8/LC_MESSAGES/messages.po [hunk 14]
```diff
@@ -16794,3 +16794,36 @@ msgstr "Modo avanzado"
...
+# msgid "Edit resource access rule"
+# msgstr ""
+
+# msgid "Rule properties"
+# msgstr ""
+
+# msgid "Resource selection"
+# msgstr ""
+
+# msgid "Contacts / Contact groups"
```
Comment (Thebarda) [inline centreon/lang/es_ES.UTF-8/LC_MESSAGES/messages.po:16810]:
```suggestion
# msgid "Contacts/Contact groups"
```

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+    return mutateAsync(payload)
+      .then((response) => {
+        const { isError } = response as ResponseError;
+        if (isError) {
+          return;
+        }
+        showSuccessMessage(t(labelMessage));
+        setModalState({ isOpen: false, mode: ModalMode.Create });
+        queryClient.invalidateQueries({ queryKey: ['resource-access-rules'] });
+      })
```
Comment (Thebarda) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts:63]: Move this code into the `onSuccess` callback in the `useMutationQuery`

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+      .finally(() => setSubmitting(false));
```
Comment (Thebarda) [inline centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts:64]: Move the code into the `onSettled`

**Review by Thebarda — DISMISSED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 17]
```diff
@@ -18593,3 +18593,36 @@ msgstr "Utilisateur supprimé"
...
 msgstr "Vous êtes sur le point de supprimer <strong>{{name}}</strong> de la liste des utilisateurs. Cet utilisateur n'aura plus accès au tableau de bord."
+
+msgid "Create a new resource access rule"
+msgstr "Créer une nouvelle règle d'accès aux ressources"
+
+msgid "Edit resource access rule"
+msgstr "Modifier une règle d'accès aux ressources"
+
+msgid "Rule properties"
+msgstr "Propriétés d'une règle"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18604]:
```suggestion
msgstr "Propriétés de la règle"
```

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 18]
```diff
@@ -18593,3 +18593,36 @@ msgstr "Utilisateur supprimé"
...
+msgstr "Sélection des ressources"
+
+msgid "Contacts / Contact groups"
+msgstr "Contacts / Groupes de contacts"
+
+msgid "The resource access rule was successfully added"
+msgstr "La règle d'accès aux ressources a été créée avec succès"
+
+msgid "The resource access rule was successfully updated"
+msgstr "La règle d'accès aux ressources a été modifée avec succès"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18616]:
```suggestion
msgstr "La règle d'accès aux ressources a été modifiée avec succès"
```

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 19]
```diff
@@ -18593,3 +18593,36 @@ msgstr "Utilisateur supprimé"
...
+msgstr "Ajouter un nouveau jeu de données"
+
+msgid "Choose at least one contact or contact group"
+msgstr "Choisissez au moins un contact ou groupe de contacts"
+
+msgid "Metaservice"
+msgstr "Méta-Service"
+
+msgid "Select resource type"
+msgstr "Sélectionnez le type de ressource"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18628]:
```suggestion
msgstr "Sélectionnez un type de ressource"
```

**Review by smau-centreon — CHANGES_REQUESTED** [review]

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 20]
```diff
@@ -18593,3 +18593,36 @@ msgstr "Utilisateur supprimé"
 
 msgid "You are going to delete <strong>{{name}}</strong> from the user list. This user will no longer access the dashboard."
```
[the remark this answers is not in the document]
Comment (dmyios) [reply centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18594]: @cg-tw it's not a part of the diff of my PR

[the remark this answers is not in the document]
Comment (cg-tw) [reply centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18594]: I know but it still needs fixing. The current label is ambiguous.

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts (hunk 15 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts]: Done

File: centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts (hunk 16 above)
Comment (dmyios) [reply centreon/www/front_src/src/ResourceAccessManagement/AddEditResourceAccessRule/Form/useFormSubmit.ts]: Done

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 21]
```diff
@@ -18594,5 +18594,38 @@ msgstr "Supprimer un utilisateur"
...
 
-msgid "You are going to delete <strong>{{name}}</strong> from the user list. This user will no longer access the dashboard."
+msgid "You are going to delete <strong>{{name}}</strong> from the user list. This user will no longer be able to access the dashboard."
 msgstr "Vous êtes sur le point de supprimer <strong>{{name}}</strong> de la liste des utilisateurs. Cet utilisateur n'aura plus accès au tableau de bord."
+
+msgid "Create a new resource access rule"
+msgstr "Créer une nouvelle règle d'accès aux ressources"
+
+msgid "Edit resource access rule"
+msgstr "Modifier une règle d'accès aux ressources"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:18604]:
```suggestion
msgstr "Modifier la règle d'accès aux ressources"
```

**Review by smau-centreon — CHANGES_REQUESTED** [review]

**Review by cg-tw — APPROVED** [review]

**Review by smau-centreon — DISMISSED** [review]

**Review by cg-tw — DISMISSED** [review]

**Review by Med1233 — DISMISSED** [review]

**Review by Thebarda — APPROVED** [review]

**Review by cg-tw — APPROVED** [review]

**Review by Thebarda — APPROVED** [review]

**Pull request merged**